### PR TITLE
2_chapter_models_yolov5_network_structure_analysis

### PR DESCRIPTION
**Pull request by ccssu**
YOLOv5 网络结构解析

### CODE REVIEW

File: document/yolov5_network_structure_analysis.md [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+# 如何理解这个depth_multiple和width_multiple呢?它决定的是整个模型中的深度（层数）和宽度（通道数）,具体怎么调整的结合后面的backbone代码解释。
+
+anchors: # 表示作用于当前特征图的Anchor大小为 xxx
+# 9个anchor，其中P表示特征图的层级，P3/8该层特征图缩放为1/8,是第3层特征
+  - [10,13, 16,30, 33,23]  # P3/8， 表示[10,13],[16,30], [33,23]3个anchor
+  - [30,61, 62,45, 59,119]  # P4/16
+  - [116,90, 156,198, 373,326]  # P5/32
+
+
+# YOLOv5 v6.0 backbone
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:29]:
```suggestion
# YOLOv5s v6.0 backbone
```

File: document/yolov5_network_structure_analysis.md [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+   [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
+   [-1, 6, C3, [256]],
+   [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
+   [-1, 9, C3, [512]],
+   [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
+   [-1, 3, C3, [1024]],
+   [-1, 1, SPPF, [1024, 5]],  # 9
+  ]
+
+# YOLOv5 v6.0 head
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:44]:
```suggestion
# YOLOv5s v6.0 head
```

File: document/yolov5_network_structure_analysis.md [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+        self.bn = nn.BatchNorm2d(c2)
+        self.act = nn.SiLU() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())
+
+    def forward(self, x):
+        return self.act(self.bn(self.conv(x)))
+
+    def forward_fuse(self, x):
+        return self.act(self.conv(x))
+```
+比如上面把width_multiple设置为了0.5，那么第一个 [64, 6, 2, 2] 就会被解析为 [3,64*0.5=32,6,2,2]，其中 第一 3为输入channel(因为输入)，32为输出channel。
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:114]:
```suggestion
比如上面把width_multiple设置为了0.5，那么第一个 [64, 6, 2, 2] 就会被解析为 [3,64*0.5=32,6,2,2]，其中第一个 3 为输入channel(因为输入)，32 为输出channel。 
```

File: document/yolov5_network_structure_analysis.md [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+<p align="center">
+  <img src="https://oneflow-static.oss-cn-beijing.aliyuncs.com/one-yolo/imgs/yolovs%E7%BD%91%E7%BB%9C%E7%BB%93%E6%9E%84%E6%A8%A1%E5%9E%8B.drawio.png" >
+  <caption> <u>图2.2</u>:yolov5s 网络整体结构 <br> </caption>
+</p>
+
+1. 详细的网络结构图：https://oneflow-static.oss-cn-beijing.aliyuncs.com/one-yolo/imgs/yolov5s.onnx.png 
+通过export.py导出的onnx格式，并通过 https://netron.app/ 网站导出的图片(模型导出 将在第八章介绍)。
+
+
+2. 模块组件右边参数 表示图片的的形状，比如 在 第 一 层( Conv )输入 图片形状 [ 3, 640, 640] ,关于这些参数，可以固定一张图片输入到网络并通过[yolov5s.yaml](https://github.com/Oneflow-Inc/one-yolov5/blob/main/models/yolov5s.yaml)的文件参数计算得到，并且可以在工程 models/[yolo.py](https://github.com/Oneflow-Inc/one-yolov5/blob/main/models/yolo.py) 通过代码进行print查看,详细数据可以参考附件表2.1。
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:163]:
```suggestion
2. 模块组件右边参数 表示特征图的的形状，比如 在 第 一 层( Conv )输入 图片形状为 [ 3, 640, 640] ,关于这些参数，可以固定一张图片输入到网络并通过[yolov5s.yaml](https://github.com/Oneflow-Inc/one-yolov5/blob/main/models/yolov5s.yaml)的模型参数计算得到，并且可以在工程 models/[yolo.py](https://github.com/Oneflow-Inc/one-yolov5/blob/main/models/yolo.py) 通过代码进行print查看,详细数据可以参考附件表2.1。
```

File: document/yolov5_network_structure_analysis.md [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+        if m in (Conv, GhostConv, Bottleneck, GhostBottleneck, SPP, SPPF, DWConv, MixConv2d, Focus, CrossConv,
+                 BottleneckCSP, C3, C3TR, C3SPP, C3Ghost, nn.ConvTranspose2d, DWConvTranspose2d, C3x):
+            # c1: 输入通道数 c2：输出通道数
+            c1, c2 = ch[f], args[0] 
+            # 该层不是最后一层，则将通道数乘以宽度因子 也就是说，宽度因子作用于除了最后一层之外的所有层
+            if c2 != no:  # if not output
+                # make_divisible的作用，使得原始的通道数乘以宽度因子之后取整到8的倍数，这样处理一般是让模型的并行性和推理性能更好。
+                c2 = make_divisible(c2 * gw, 8)
+
+            # 将前面的运算结果保存在args中，它也就是最终的方法参数。
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:219]:
```suggestion
            # 将前面的运算结果保存在args中，它也就是这个模块最终的输入参数。
```

File: document/yolov5_network_structure_analysis.md [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+        y = []  # outputs
+        for si, fi in zip(s, f):
+            xi = scale_img(x.flip(fi) if fi else x, si, gs=int(self.stride.max()))
+            yi = self._forward_once(xi)[0]  # forward
+            # cv2.imwrite(f'img_{si}.jpg', 255 * xi[0].cpu().numpy().transpose((1, 2, 0))[:, :, ::-1])  # save
+            yi = self._descale_pred(yi, fi, si, img_size)
+            y.append(yi)
+        y = self._clip_augmented(y)  # clip augmented tails
+        return flow.cat(y, 1), None  # augmented inference, train
+    # 前向传播
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:326]:
```suggestion
    # 前向传播具体实现
```

File: document/yolov5_network_structure_analysis.md [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+        else:
+            x, y, wh = p[..., 0:1] / scale, p[..., 1:2] / scale, p[..., 2:4] / scale  # de-scale
+            if flips == 2:
+                y = img_size[0] - y  # de-flip ud
+            elif flips == 3:
+                x = img_size[1] - x  # de-flip lr
+            p = flow.cat((x, y, wh, p[..., 4:]), -1)
+        return p
+    
+    def _clip_augmented(self, y):
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:373]: 这个是TTA的时候对原图片进行裁剪，也是一种数据增强方式，用在TTA测试的时候。

File: document/yolov5_network_structure_analysis.md [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+        m = self.model[-1]  # Detect() module
+        for mi, s in zip(m.m, m.stride):  # from
+            b = mi.bias.view(m.na, -1).detach()  # conv.bias(255) to (3,85)
+            b[:, 4] += math.log(8 / (640 / s) ** 2)  # obj (8 objects per 640 image)
+            b[:, 5:] += math.log(0.6 / (m.nc - 0.999999)) if cf is None else flow.log(cf / cf.sum())  # cls
+            mi.bias = flow.nn.Parameter(b.view(-1), requires_grad=True)
+    #  打印模型中最后Detect层的偏置biases信息(也可以任选哪些层biases信息)
+    def _print_biases(self):
+        """
+        打印模型中最后Detect层的偏置biases信息(也可以任选哪些层biases信息)
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:409]:
```suggestion
        打印模型中最后Detect模块里面的卷积层的偏置biases信息(也可以任选哪些层biases信息)
```

Comment (BBuf) [reply document/yolov5_network_structure_analysis.md:409]: 这里的也可以任选哪些层指的是什么意思？

File: document/yolov5_network_structure_analysis.md [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+            if isinstance(m.anchor_grid, list):
+                m.anchor_grid = list(map(fn, m.anchor_grid))
+        return self
+```
+
+###  Detect类解读
+```python
+class Detect(nn.Module):
+    """
+    Detect模块是用来构建Detect层的，将输入feature map 通过一个卷积操作和公式计算到我们想要的shape, 为后面的计算损失或者NMS作准备
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:460]:
```suggestion
    Detect模块是用来构建Detect层的，将输入feature map 通过一个卷积操作和公式计算到我们想要的shape, 为后面的计算损失或者NMS后处理作准备
```

File: document/yolov5_network_structure_analysis.md [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+        z = []  # inference output
+        for i in range(self.nl):
+            x[i] = self.m[i](x[i])  # conv
+            bs, _, ny, nx = x[i].shape  # x(bs,255,20,20) to x(bs,3,20,20,85)
+            x[i] = x[i].view(bs, self.na, self.no, ny, nx).permute(0, 1, 3, 4, 2).contiguous()
+
+            if not self.training:  # inference
+                if self.onnx_dynamic or self.grid[i].shape[2:4] != x[i].shape[2:4]:
+                    self.grid[i], self.anchor_grid[i] = self._make_grid(nx, ny, i)
+                # 	向前传播时需要将相对坐标转换到grid绝对坐标系中
```
Comment (BBuf) [inline document/yolov5_network_structure_analysis.md:495]: 这一行有点错位

**Review by BBuf — APPROVED** [review]
修改完上面的所有comment可以合并

**Review by ccssu — COMMENTED** [review]
处理格式

**Review by BBuf — APPROVED** [review]
